spinalcordtoolbox/image.py

```diff
@@ -401,6 +404,25 @@ def set_qform_to_sform(self):
         self.hdr.set_qform(self.hdr.get_sform())
         self.hdr._structarr['qform_code'] = self.hdr._structarr['sform_code']
 
+    def fix_header_dtype(self, header='self', array='self'):
+        """
+        Change the header dtype to the match the datatype of the array.
+
+        Note: This method defaults to operating on the image's `data` and `hdr` properties,
+              but the fix can also be applied to supplied arrays and headers, too."""
+        hdr = self.hdr if header == 'self' else header
+        data = self.data if array == 'self' else array
```

Comment:
> They are actually doing the same thing, aren't they? Can they not be replaced by a single function/method to reduce code duplications?

> I agree, a method like `self.fix_header_dtype()` called from both places would make sense.

I've done this in [65e9b31](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3944/commits/65e9b313f47039f1d94c1a7ce2bd80453afc7416). The tricky thing, though, was that:

- The `load` LOCs acted on the properties of the image itself (so, a `fix_header_dtype` method could act on `self.data` and `self.hdr`)
- The `save` LOCs acted on _copies_ of the properties, meaning that `self.data` and `self.hdr` were off-limits.

To account for both cases while still using a class method, I had to use a sort of awkward pattern. So, we may also want to consider making this an external function instead? I'm not sure... :thinking:

Reply:
I decided to refactor `Image.save()` so that `fix_header_dtype()` can be simplified, I hope that's ok.